proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+   `xds.shard_id=1234` and potentially change through the lifetime of an xDS
+   stream.
+
+3. Context parameters from the URL are added next, in the above example
+   `xds.bar=baz`.
+
+4. [Client feature
+   capability](https://www.envoyproxy.io/docs/envoy/latest/api/client_features)
+   context parameters are then added, e.g. clusters at the client might have
+   `{xds.client_feature.lb.least_loaded: true}`. Client feature capabilities
+   will be required to be declared with respect to a specific resource type URL.
+   All these attributes will be `xds.client_feature.` prefixed.
+
+5. Finally per-resource type well-known attributes are added, e.g. an on-demand
+   listener load might have `{xds.resource.vip: 96.54.3.1}`. These attributes
+   are generated by code, e.g. an on-demand LDS implementation will first detect
+   the VIP, halt the request, ask for the `xds.resource.vip` from the control
+   plane, and then resume the request after discovery is completed. All these
+   attributes will be `xdstp.resource.` prefixed.
+
+An example computed URN following the above example is
+`xdstp://some-authority/some.type/foo?xds.node.metadata.foo=bar&xds.shard_id=1234&xds.bar=baz&xds.client_feature.lb.least_loaded=false&xds.resource.vip=96.54.3.1`.
+
+This proposal reserves the `@` prefix for context parameter values for future
+URI context parameter enhancements.
+
+### Collections
+
+Two forms of resources collections are described below, *list* and *glob*. These
+different forms reflect the tension between the need for explicit collection
+representation (motivating list collections) and scalability concerns as this
+explicit representation becomes a botleneck (motivating glob collections).
+
+#### List
+
+List collections provide a collection directory as a first-class xDS resource.
+For a given xDS resource type `<T>` the list collection type is named by
+convention as `<T>Collection` and defined as:
+
+```
+message <T>Collection {
+  repeated CollectionEntry resources = 1;
+}
+```
+
+where `CollectionEntry` is defined:
+
+
+```
+message CollectionEntry {
+  // Inlined resource entry.
+  message InlineEntry {
+    // Optional name to describe the inlined resource. Resource names must
+    // [a-zA-Z0-9_-\./]+ (TODO(htuch): turn this into a PGV constraint once
+    // finalized, probably should be a RFC3986 pchar). This name allows
+    // reference via the #entry directive in ResourceLocator.
+    string name = 1 [(validate.rules).string.pattern = "^[0-9a-zA-Z_\\-\\.~:]+$"];
+
+    // The resource's logical version. It is illegal to have the same named xDS
+    // resource name at a given version with different resource payloads.
+    string version = 2;
+
+    // The resource payload, including type URL.
+    google.protobuf.Any resource = 3;
+  }
+
+  oneof resource_specifier {
+    option (validate.required) = true;
+
+    // A resource locator describing how the member resource is to be located.
+    ResourceLocator locator = 1;
```

Comment:
Should this just be a `string`, considering that we're using the string form everywhere else?  (As mentioned above, I know that we've been discussing introducing this structured form for wire-efficiency reasons, but I assume that will be covered in a subsequent xRFC.)